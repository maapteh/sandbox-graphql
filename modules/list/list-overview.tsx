import React, { useState } from 'react';
import styled from 'styled-components';
import { useListsQuery } from '../../codegen/_graphql';
import { ListItems } from './list-items';

const List = styled.div`
    padding: 8px;
    margin: 0 0 24px;
    border: 1px solid #eee;
`;

const START_INITIAL = 0;
const PAGE_SIZE = 2;

export const ListOverview: React.FC = () => {
    const [start, setStart] = useState(START_INITIAL);
    const { loading, error, data, fetchMore } = useListsQuery({
        variables: {
            start: START_INITIAL,
            size: PAGE_SIZE,
        },
    });

    if (loading) {
        return <p>Loading...</p>;
    }

    if (error) {
        return <p>Error: {error}</p>;
    }

    return (
        <>
            {data?.lists?.result?.map((list) => (
                <List key={list.id}>
                    {list.description}
                    <ListItems id={list.id} />
                </List>
            ))}
            <button
                onClick={() => {
                    const newStart = start + PAGE_SIZE;
                    if (data?.lists?.total && newStart < data.lists.total) {
                        fetchMore({
                            variables: {
                                start: newStart,
                                size: PAGE_SIZE,
                            },
                            updateQuery: (
                                previousResult,
                                { fetchMoreResult },
                            ) => {
                                if (!fetchMoreResult?.lists) {
                                    return previousResult;
                                }

                                return {
                                    ...previousResult,
                                    lists: {
                                        ...fetchMoreResult.lists,
                                        total: fetchMoreResult.lists.total,
                                        result: [
                                            ...(previousResult.lists?.result ||
                                                []),
                                            ...(fetchMoreResult.lists.result ||
                                                []),
                                        ],
                                    },
                                };
                            },
                        });
                        setStart(newStart);
                    }
                }}
            >
                Load More
            </button>
        </>
    );
};
