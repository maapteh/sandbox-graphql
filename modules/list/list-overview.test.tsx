import React from 'react';
import { MockedProvider, wait } from '@apollo/react-testing';
import { render, cleanup, act } from '@testing-library/react';
import { InMemoryCache } from 'apollo-cache-inmemory';
import { fragmentMatcher } from '../../lib/fragment-matcher';
import { ListOverview } from './list-overview';
import {
    MOCK_LISTS_FOUND,
    MOCK_LISTS_NOT_FOUND,
    MOCK_LISTS_ERROR,
} from './__mocks__/list-mocks';

describe('lists', () => {
    afterEach(cleanup);

    it('renders when we have data', async () => {
        const { container } = render(
            <MockedProvider
                mocks={MOCK_LISTS_FOUND}
                addTypename
                cache={
                    new InMemoryCache({
                        addTypename: true,
                        fragmentMatcher,
                    })
                }
            >
                <ListOverview />
            </MockedProvider>,
        );
        await act(async () => await wait(0));

        expect(container).toMatchSnapshot();
    });

    it('shows blank without data', async () => {
        const { container } = render(
            <MockedProvider
                mocks={MOCK_LISTS_NOT_FOUND}
                addTypename
                cache={
                    new InMemoryCache({
                        addTypename: true,
                        fragmentMatcher,
                    })
                }
            >
                <ListOverview />
            </MockedProvider>,
        );
        await act(async () => await wait(0));

        expect(container).toMatchSnapshot();
    });

    it('shows error', async () => {
        const { container } = render(
            <MockedProvider
                mocks={MOCK_LISTS_ERROR}
                addTypename
                cache={
                    new InMemoryCache({
                        addTypename: true,
                        fragmentMatcher,
                    })
                }
            >
                <ListOverview />
            </MockedProvider>,
        );
        await act(async () => await wait(0));

        expect(container).toMatchSnapshot();
    });
});
