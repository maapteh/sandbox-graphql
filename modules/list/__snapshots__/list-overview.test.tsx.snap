// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`lists renders when we have data 1`] = `
<div>
  <div
    class="sc-AxhCb dtlpSM"
    data-testid="list-1"
  >
    list-1
    <a
      class="sc-AxhUy cmnYOP"
    >
      rename
    </a>
    <div
      class="sc-AxjAm dnFUoP"
    />
  </div>
  <div
    class="sc-AxhCb dtlpSM"
    data-testid="list-2"
  >
    list-2
    <a
      class="sc-AxhUy cmnYOP"
    >
      rename
    </a>
    <div
      class="sc-AxjAm dnFUoP"
    />
  </div>
  <button>
    Load More
  </button>
</div>
`;

exports[`lists shows blank without data 1`] = `
<div>
  <button>
    Load More
  </button>
</div>
`;

exports[`lists shows error 1`] = `
<div>
  <p
    data-testid="error"
  >
    Error 
    {"graphQLErrors":[],"networkError":{},"message":"Network error: big error"}
  </p>
</div>
`;
